File: node-martian-ddd/src/domain/robot.js
"use strict";
const { left, right, step } = require("./orientation");

class Robot {
  constructor(position, orientation) {
    this.position = position;
    this.orientation = orientation;
    this.lost = false;
  }
  turnLeft() {
    this.orientation = left(this.orientation);
  }
  turnRight() {
    this.orientation = right(this.orientation);
  }
  nextForward() {
    const [dx, dy] = step(this.orientation);
    return { x: this.position.x + dx, y: this.position.y + dy };
  }
}

module.exports = { Robot };

File: go-martian-ddd/internal/domain/robot.go
package domain

type Robot struct {
	Position    Position
	Orientation Orientation
	Lost        bool
}

func (r *Robot) TurnLeft()  { r.Orientation = r.Orientation.Left() }
func (r *Robot) TurnRight() { r.Orientation = r.Orientation.Right() }

func (r *Robot) NextForward() Position {
	dx, dy := r.Orientation.Step()
	return Position{X: r.Position.X + dx, Y: r.Position.Y + dy}
}
